character-controller.js

```diff
@@ -1237,6 +1237,8 @@ class NpcPlayer extends StaticUninterpolatedPlayer {
       visemes: true,
       debug: false,
     });
+    avatar.app = app;
+    this.appManager.apps.push(app); // note: for `physicsManager.simulatePhysics` can correctly update ragdoll's bones from physx result.

```

Comment:
I don't think this is valid. The character controller should not be adding apps to the app manager.

Reply:
If this is needed and was not added then we should understand why that is the case. It's not the character controller's job to fix it.

Reply:
The app should either be transplanted from another appManager, or it should be explicitly added to the ownership of this npc. But to make setting the avatar change app ownership like this is probably not correct.

Reply:
I'll try to use another proper way.

Reply:
Hello @avaer, I aligned `NpcPlayer` with `LocalPlayer`, added `setAvatarUrl()` function, which will push `app` to `appManeger` automatically. Is it proper?
Sibling commit: https://github.com/webaverse/npc/pull/4/commits/30fcbae4a7820dd87dad8cd1b5e7bf53260212de

I'll keep trying to avoid `avatar.app = app;` in `character-controller.js`.

Reply:
Found that `localPlayer` also set `avatar.app = app` immediately after `new Avatar` : https://github.com/webaverse/app/blob/01407552c6e28067c0af346ca5b0e6d70646191e/player-avatar-binding.js#L61

So I set `npcPlayer`'s `avatar.app = app` immediately after `new Avatar` is proper already? https://github.com/gonnavis/app/blob/77a79ce31e77d950120ff0f0d621e9e2a4d2481c/character-controller.js#L1244

Reply:
No this doesn't work. We already have the VRM app by the time we call `createNpc`. We should not load it again. Also, that app is probably already owned by some app manager. You can't add it to another app manager too.

https://github.com/webaverse/app/blob/f6eb31d3467076946b625381b7b4138a70a5953d/metaverse_components/npc.js#L56

I think what you are missing is update() function calling for this app or NPC? If you need that delegation you can do it in metaverse_components/npc.js, which owns the NPC.

Reply:
@avaer Now I add them in `npc.js`, is it ok?
https://github.com/webaverse/app/pull/2731/commits/fcb6b4ccd18066de36935985551a611f0f7bf9d9#diff-4cab899c225f756cb4eee00c66e53f8a2cd4399d86b0ae3d87d8d3d93919c68fR64